### ISSUE TITLE
Saving Dataframe for post processing and adding version info

### ISSUE THREAD

**Issue by itsdfish** [StatisticalRethinkingJulia/MCMCBenchmarks.jl#10]
Because some of the benchmarking takes hours to generate, I was thinking it might be a good idea to save the results dataframe for post-processing. It might be a good idea to include version info such as Julia, Stan, package versions, and hardware specs. This would allow us to track changes across versions and reproduce the results if necessary. 

I'm assuming there might be a way to extract most of this information with functions rather than entering it into a dataframe manually. Any thoughts?

**Comment by goedman** [comment]
Pkg.installed() gives a Dict back listing all installed packages. `] st DynamicHMC` returns a much smaller, better list. Manifest.toml is kind of covering some of this, we could save that file. For CmdStan it's in first couple of lines of the .csv files (or I could make a small extract_version() function).

I just added some of that (saving the chains) to lr.jl. Unfortunately I don't think this is portable, although I have never done that with DataFrames, maybe that works better.

Particularly for the long running simulations I do like to save multiple chains and be able to go back and restore these chains. I wonder if the best solution is to store these somewhere outside .julia using a similar structure as Examples but as a subdirectory of a couple of time stamped directories with configuration and hardware specific info? 

The same for the plots, right now PyPlot crashes Julia on my system, that's why I use GR. Maybe it's time to reinstall all of Julia once more.

Anyway, working through a relative simple example (LinearRegression) I'll also run into many of your questions.

**Comment by itsdfish** [comment]
I'm definitely open to suggestions, but I like the idea of saving version metadata as a csv in the results folder or some other logical location. The metadata and benchmark data could be linked by a unique run ID listed in the file name (or unique subfolder). If we want to compare versions of a package, it would be easy to load the CSVs as a dataframes, add metadata and concatenate the dataframes. The version info might look like this:

| Software | Version      |
|----------|--------------|
| Julia    | 1.1.1        |
| Stan     | 2.19         |
| Turing   | 0.6.17       |
| OS       | Ubuntu 16.04 |
...

As you noted, getting package information will be easy. I'm not sure about the Julia version information. `versioninfo` only seems to print to the REPL, but I could be wrong. I'm not sure about system information. I think there are terminal commands but I'm not sure whether its possible to toggle back and fort between REPL and terminal using code as opposed to manually entered commands.

**Comment by itsdfish** [comment]
Just to clarify, are you thinking about saving each chain in the benchmark?

I was thinking of only saving the results DataFrame used to construct the plots because storing the chains would be a lot of data.

**Comment by goedman** [comment]
Absolutely, in the benchmark package we’ll only save the benchmark data and the plots in pyplot format.

**Comment by goedman** [comment]
Hi Chris,

When doing the LinearRegression example I can also look at inserting the dHMC stepsize.

2 quick ideas:

1. The benchmark() call, as in the model files, seems the same in all models, should we move it the the src directory?

2. Getting PyPlot to work on my system is a failure, so I'l generate a .csv file for the Gaussian example and send it to you. That should be portable.

Rob

**Comment by itsdfish** [comment]
Yeah. I think that's a good idea to use a common benchmark function. Currently, only the number of data points is varied. So that makes sense.  If we want to vary other aspects of the simulation, we can create simulation-specific version of benchmark for that purpose.

**Comment by goedman** [comment]
For the LR example I also vary the number of parameters/coefficients - size(X, 2) - and I wonder if we should pass in the number of chains? I haven't quite figured out how to do that for hierarchical models (m12.6) but one step at the time.

**Comment by itsdfish** [comment]
You can simply pass the number of chains as a keyword argument if you want it to be constant and catch it with the `modifyConfig!` function. I would recommend setting a default value `Nchains=1` in `modifyConfig` so its possible to override if desired.  By contrast, if you want to vary the number of chains within the benchmark simulation, you will have to create a new benchmark function with a for loop for the chains. However, varying the number of chains as a factor may not be very interesting.

I considered using 1 chain to reduce the demand for computational resources on my system. Another benefit is that it also leaves open the possibility of running each sampler on a separate processor:

`results = pmap(s->benchmark(s,...),samplers)`

which would be easy to implement for longer running simulations. 

In terms of number of coefficients, I recommend passing that as a keyword argument:

e.g.` myModel(;Ncoefs=Ncoefs,...,kwargs...)`

When `Ncoefs` is passed to other functions, such as `modifyConfig!`, it will be ignored in `kwargs`. But the values will be recorded in results dataframe via the function `addKW!`.

**Comment by itsdfish** [comment]
I guess you would have to set Nchains in `runSampler`. Approximately like:

```
function runSampler(s::AHMCNUTS,data;Nchains=1,kwargs...)
    Turing.turnprogress(false)
    return pmap(x->sample(s.model(data...),1:Nchains,s.config)
end
```
But with the `reducemap` call. The more I think about it, it seems like running 

`results = pmap(s->benchmark(s,...),samplers)`

would be better because it would speed things up more and the number of chains may not be an interesting factor to vary.

**Comment by goedman** [comment]
Ok, let’s go for single chain per sampler. With the pmap() option, would timing still be accurate?

Just merged a results.csv file for the Gaussian example up to 2000 observations. The gr() plots look ok, but if you want to generate the pyplot plots, that’s fine as well.

**Comment by itsdfish** [comment]
Good question. As far as I know, time measurements will not be affected because the` @time` macro is in the `benchmark!` function. So the overhead associated with loading the variables and functions should not be included in the timing. I don't think there is any reason the processors should be different, unless some sort of competition over the L caches occurs when running programs in parallel. My inner scientist says that we could run the Gaussian model both ways and see if the results are similar. 

The gr plots aren't too bad actually. Maybe I will generate pyplot figures once the entire suite is working.

**Comment by itsdfish** [comment]
I added the ability to save the results dataframe as a csv in a time stamped subfolder of results. I will add metadata containing package info.

**Comment by itsdfish** [comment]
It is now possible to save a seperate metadata file containing system info and software info.

**Comment by goedman** [comment]
Nice Chris! Do you want to register the package or would you prefer me to do it? It'll take 3 days as it is a new package.

**Comment by itsdfish** [comment]
Thanks Rob. Registering the project sounds like a good idea. Would you mind registering so I can see how you do it?

P.S. I am about to push some changes to master that simplify the benchmark configuration. Previously, each time we want to change the benchmark configuration,  we had to define a new `benchmark` and `pbenchmark` function to loop  through the relevant variables. Now this can be done with a single function. Suppose you want to benchmark with data sample size 10 and 50. This would be written like this:

```
options = (Nsamples=2000,Nadapt=1000,delta=.8,Nd=[10,50])
results = pbenchmark(samplers,simulateLBA,Nreps;options...)

```
Now suppose you want to also vary delta. This can be accomplished like so:

```
options = (Nsamples=2000,Nadapt=1000,delta=[.6,.8],Nd=[10,50])
results = pbenchmark(samplers,simulateLBA,Nreps;options...)
```
My solution was to define a lazy iterator that creates all of the permutations of variables listed in options, thus emulating a variable number of for loops. I like this solution quite well because it is very flexible and no longer requires different versions of `benchmark` and `pbenchmark`.

**Comment by itsdfish** [comment]
By the way, do you know how to get a dictionary of all installed packages and dependencies? The current setup crashes now because AdvancedHMC is a dependency instead of an "installed" package. The closest thing I could find was `Pkg.status(mode=PKGMODE_MANIFEST)`, but that only prints to the console.

**Comment by goedman** [comment]
Isn’t that just printing the environment Manifest.toml file ? E.g. under .julia/Environments/1.1 if you are in Julia 1.1?

**Comment by goedman** [comment]
I’ll register later today. It should really just be opening a new issue.

**Comment by itsdfish** [comment]
Yeah. Exactly. It prints out the Manifest.toml file to the REPL.  Ideally, we could extract that information from a dictionary, like Pkg.installed(), but I couldn't figure out a solution. I suppose we could save the Manifest.toml file and define a method for parsing it during post processing. Does that sound reasonable?

**Comment by itsdfish** [comment]
I am converging on a solution. This will return a dictionary with all packages:

```
using TOML
d=TOML.parsefile("/home/dfish/.julia/environments/v1.1/Manifest.toml")
```
However, I'm not sure how to include TOML as a dependency because it is an unregistered package. The other problem is that I cannot figure out how to access Manifest.toml on any machine. Any ideas?

**Comment by goedman** [comment]
I saw some messages suggesting:
```
Pkg.clone("https://github.com/wildart/TOML.jl")
```
for now. Apparently a clean solution will follow later (after Pkg3 is a bit further along).

I would for now use ```userdir = expanduser("~")``` to get the first part:
```
d=TOML.parsefile(joinpath(userdir, ".julia/environments/v1.1/Manifest.toml"))
```
Was that your question?

**Comment by itsdfish** [comment]
Thanks Rob. That gets me much closer. The last complication is that we cannot always assume `.../v1.1/...` will be the correct environment version. I've been searching all around, including the ENV dictionary of paths, but to no avail. 

After running ]update, the following message are printed:

```
(v1.1) pkg> update
  Updating registry at `~/.julia/registries/General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
  Updating git-repo `https://github.com/wildart/TOML.jl.git`
 Resolving package versions...
 Installed Tables ─ v0.2.5
  Updating `~/.julia/environments/v1.1/Project.toml`
 [no changes]
  Updating `~/.julia/environments/v1.1/Manifest.toml`
```
The last line suggests the exact path can be obtained somehow. Perhaps in the absence of an elegant solution I could parse the folder name from:

```
 VERSION
  v"1.1.1"
```

I suppose the rule would be to change v"1.1.1" to "v1.1".

**Comment by itsdfish** [comment]
Well, I have a stopgap solution for the time being. If you have something more elegant, I will add that instead. 

```
function getpath()
    userdir = expanduser("~")
    str = split(string(VERSION),".")[1:end-1]
    str1 = [str[s]*"." for s in 1:length(str)-1]
    push!(str1,str[end])
    version = string(str1...)
    path = joinpath(userdir, string(".julia/environments/v",version,"/Manifest.toml"))
    return path
end
```
That should work as long as the versioning format does not change. 


The last issue is removing the following warning message:

```
using MCMCBenchmarks
[ Info: Recompiling stale cache file /home/dfish/.julia/compiled/v1.1/MCMCBenchmarks/xNtfs.ji for MCMCBenchmarks [72ce6e51-3f7f-5ebd-8e88-b86aa3e31c30]
  Updating registry at `~/.julia/registries/General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
  Updating git-repo `https://github.com/wildart/TOML.jl`
 Resolving package versions...
  Updating `~/.julia/environments/v1.1/Project.toml`
  [191fdcea] ~ TOML v0.4.0 [`~/.julia/dev/TOML`] ⇒ v0.4.0 #master (https://github.com/wildart/TOML.jl)
  Updating `~/.julia/environments/v1.1/Manifest.toml`
  [191fdcea] ~ TOML v0.4.0 [`~/.julia/dev/TOML`] ⇒ v0.4.0 #master (https://github.com/wildart/TOML.jl)
┌ Warning: Package MCMCBenchmarks does not have TOML in its dependencies:
│ - If you have MCMCBenchmarks checked out for development and have
│   added TOML as a dependency but haven't updated your primary
│   environment's manifest file, try `Pkg.resolve()`.
│ - Otherwise you may need to report an issue with MCMCBenchmarks
└ Loading TOML into MCMCBenchmarks from project dependency, future warnings for MCMCBenchmarks are suppressed.
```
I'm not sure how to add TOML to the Manifest.toml file. Here is what happens:

```
(v1.1) pkg> activate MCMCBenchmarks

(MCMCBenchmarks) pkg> add TOML
 Resolving package versions...
ERROR: Unsatisfiable requirements detected for package TOML [037cace4]:
 TOML [037cace4] log:
 ├─possible versions are: [0.1.0-0.1.2, 0.2.0, 0.3.0] or uninstalled
 ├─restricted to versions * by an explicit requirement, leaving only versions [0.1.0-0.1.2, 0.2.0, 0.3.0]
 └─restricted by julia compatibility requirements to versions: uninstalled — no versions left
```

**Comment by itsdfish** [comment]
Well, it seems like the easiest thing to do is to add the TOML.jl files to our package. Based on discourse conversations, there does not appear to be an elegant solution. 

If you know of one, we can use that instead. The other feature that might be improved is [defining](https://github.com/StatisticalRethinkingJulia/MCMCBenchmarks.jl/issues/10#issuecomment-497923241) the path to the Manifest.toml file.  For the time being, I will close this for now since these hacks seem to work.
